TYPO3.Neos/Classes/TYPO3/Neos/TypoScript/DimensionsMenuImplementation.php

```diff
@@ -0,0 +1,216 @@
+ * - labelExpression (string): Expression to use for label generation if "dimension" is not set.
+ */
+class DimensionsMenuImplementation extends AbstractMenuImplementation
+{
+
+    /**
+     * @Flow\Inject
+     * @var ConfigurationContentDimensionPresetSource
+     */
+    protected $configurationContentDimensionPresetSource;
+
+    /**
+     * @Flow\Inject
+     * @var ContentDimensionCombinator
+     */
+    protected $contentDimensionCombinator;
+
+    /**
+     * @Flow\Inject(lazy=false)
+     * @var \TYPO3\Eel\EelEvaluatorInterface
+     */
+    protected $eelEvaluator;
+
+    /**
+     * @Flow\InjectConfiguration(package="TYPO3.TYPO3CR", path="labelGenerator.eel.defaultContext")
+     * @var array
+     */
+    protected $defaultContextConfiguration;
+
+    /**
+     * @return string
+     */
+    public function getDimension()
+    {
+        return $this->tsValue('dimension');
+    }
+
+    /**
+     * @return array
+     */
+    public function getPresets()
+    {
+        return $this->tsValue('presets');
+    }
+
+    /**
+     * @return array
+     */
+    public function getIncludeAllPresets()
+    {
+        return $this->tsValue('includeAllPresets');
+    }
+
+    /**
+     * @return string
+     */
+    public function getLabelExpression()
+    {
+        return $this->tsValue('labelExpression');
+    }
+
+    /**
+     * Builds the array of Menu items for this variant menu
+     */
+    protected function buildItems()
+    {
+        $menuItems = [];
+        $targetDimensionsToMatch = [];
+        $allDimensionPresets = $this->configurationContentDimensionPresetSource->getAllPresets();
+        $includeAllPresets = $this->getIncludeAllPresets();
+        $pinnedDimensionValues = $this->getPresets();
+
+        $pinnedDimensionName = $this->getDimension();
+        if ($pinnedDimensionName !== null) {
+            $targetDimensionsToMatch = $this->currentNode->getContext()->getTargetDimensions();
+            unset($targetDimensionsToMatch[$pinnedDimensionName]);
+        }
+
+        foreach ($this->contentDimensionCombinator->getAllAllowedCombinations() as $allowedCombination) {
+            $targetDimensions = $this->calculateTargetDimensionsForCombination($allowedCombination);
+
+            if ($pinnedDimensionName !== null && is_array($pinnedDimensionValues)) {
+                if (!in_array($targetDimensions[$pinnedDimensionName], $pinnedDimensionValues)) {
+                    continue;
+                }
+            }
+
+            // skip variants not matching the current target dimensions (except the dimension this menu covers)
+            if ($targetDimensionsToMatch !== []) {
+                foreach ($targetDimensionsToMatch as $dimensionName => $dimensionValue) {
+                    if ($targetDimensions[$dimensionName] !== $dimensionValue) {
+                        continue 2;
+                    }
+                }
+            }
```

Comment:
This whole block can be removed if the handling of the `dimension` was set case is moved to a separate method which iterates over the result of `\TYPO3\TYPO3CR\Domain\Service\ContentDimensionPresetSourceInterface::getAllowedDimensionPresetsAccordingToPreselection` instead of all combinations.